ocaml/xapi/cert_refresh.ml

```diff
@@ -0,0 +1,102 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct let name = "cert_refresh" end)
+
+open D
+
+let replace_extension filename ~ext =
+  let base =
+    match Filename.extension filename with
+    | "" ->
+        filename
+    | _ ->
+        Filename.remove_extension filename
+  in
+  Printf.sprintf "%s.%s" base ext
+
+let cert_path = function
+  | `host ->
+      !Xapi_globs.server_cert_path
+  | `host_internal ->
+      !Xapi_globs.server_cert_internal_path
+
+let new_cert_path type' = replace_extension (cert_path type') ~ext:"new"
+
+let backup_cert_path type' = replace_extension (cert_path type') ~ext:"bak"
+
+let new_host_cert ~dbg ~path : X509.Certificate.t =
+  let name, ip =
+    match Networking_info.get_management_ip_addr ~dbg with
+    | None ->
+        let msg = Printf.sprintf "%s: failed to get management IP" __LOC__ in
+        D.error "%s" msg ;
+        raise Api_errors.(Server_error (internal_error, [msg]))
+    | Some ip ->
+        ip
+  in
+  let dns_names = Networking_info.dns_names () in
+  let ips = [ip] in
+  Gencertlib.Selfcert.host ~name ~dns_names ~ips path
+
+let host ~__context ~type' =
+  let host = Helpers.get_localhost ~__context in
+  let dbg = Context.string_of_task __context in
+  let pem = cert_path type' in
+  let path = new_cert_path type' in
+  let cert = new_host_cert ~dbg ~path in
+  let bak = backup_cert_path type' in
```

Comment:
We might want to think about the case where `bak` exists on the file system at this point - it probably means that a previous cert refresh failed. Do we just error out in this case or try to resolve the problem?

Reply:
Ah I see we don't actually remove `bak` at the end. The idea is that it is useful to keep around in case the user needs to manually intervene because of a failure?

I'm thinking about the case where a cert refresh has failed - the user's first instinct after running `xe cert-refresh` and seeing a failure is going to be to run it again, so it would be nice not to overwrite `bak` in this case

Reply:
Good point! Maybe we should fail if the backup exists